src/repository.c

```diff
@@ -177,6 +177,10 @@ void git_repository_free(git_repository *repo)
 	git__free(repo->ident_name);
 	git__free(repo->ident_email);
 
+	if (repo->hook_payload != NULL && repo->hook_payload_free != NULL) {
```

Comment:
This is a bit weird to use. I'd understand not calling `hook_payload_free` if its type wasn't called `git_hook_destructor`. With that naming, though, a user might expect it to also be called in case there is no payload at all